sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala

```diff
@@ -524,23 +524,142 @@ abstract class TypedImperativeAggregate[T] extends ImperativeAggregate {
+   *
+   * This is only called when doing Partial or PartialMerge mode aggregation, before the framework
+   * shuffle out aggregate buffers.
+   */
+  def serializeAggregateBufferInPlace(buffer: InternalRow): Unit = {
+    buffer(mutableAggBufferOffset) = serialize(getBufferObject(buffer))
+  }
+}
+
+/**
+ * Aggregation function which allows **arbitrary** user-defined java object to be used as internal
+ * aggregation buffer for Hive.
+ */
+abstract class HiveTypedImperativeAggregate[T] extends TypedImperativeAggregate[T] {
+
+  /**
+   * Creates an empty aggregation buffer object for partial 1 mode. This is called
+   * before processing each key group(group by key).
+   *
+   * @return an aggregation buffer object
+   */
+  def createAggregationBuffer(): T
+
+  /**
+   * Creates an empty aggregation buffer object for partial 2 mode.
+   *
+   * @return an aggregation buffer object
+   */
+  def createPartial2ModeAggregationBuffer(): T
+
+  var partial2ModeBuffer: InternalRow = _
+
+  /**
+   * Updates the aggregation buffer object with an input row and returns a new buffer object. For
+   * performance, the function may do in-place update and return it instead of constructing new
+   * buffer object.
+   *
+   * This is typically called when doing Partial or Complete mode aggregation.
+   *
+   * @param buffer The aggregation buffer object.
+   * @param input an input row
+   */
+  def update(buffer: T, input: InternalRow): T
+
+  /**
+   * Merges an input aggregation object into aggregation buffer object and returns a new buffer
+   * object. For performance, the function may do in-place merge and return it instead of
+   * constructing new buffer object.
+   *
+   * This is typically called when doing PartialMerge or Final mode aggregation.
+   *
+   * @param buffer the aggregation buffer object used to store the aggregation result.
+   * @param input an input aggregation object. Input aggregation object can be produced by
+   *              de-serializing the partial aggregate's output from Mapper side.
+   */
+  def merge(buffer: T, input: T): T
+
+  /**
+   * Generates the final aggregation result value for current key group with the aggregation buffer
+   * object.
+   *
+   * Developer note: the only return types accepted by Spark are:
+   *   - primitive types
+   *   - InternalRow and subclasses
+   *   - ArrayData
+   *   - MapData
+   *
+   * @param buffer aggregation buffer object.
+   * @return The aggregation result of current key group
+   */
+  def eval(buffer: T): Any
+
+  /** Serializes the aggregation buffer object T to Array[Byte] */
+  def serialize(buffer: T): Array[Byte]
+
+  /** De-serializes the serialized format Array[Byte], and produces aggregation buffer object T */
+  def deserialize(storageFormat: Array[Byte]): T
+
   final override def initialize(buffer: InternalRow): Unit = {
+    partial2ModeBuffer = buffer.copy()
+    partial2ModeBuffer(mutableAggBufferOffset) = createPartial2ModeAggregationBuffer()
```

Comment:
I'm a little lost here. So this `HiveTypedImperativeAggregate` has 2 buffers? What's the difference between `partial2ModeBuffer` and `buffer`?

Reply:
So Spark Catalyst has designed UDAF execution such that it uses one aggregation buffer for performing the aggregations for all UDAF operators(Sort based, Object hash based etc.) which makes sense from Spark's point of view. However, from Hive's point of view, two aggregate buffers are expected to be used, one for PARTIAL1/COMPLETE and the other for PARTIAL2/FINAL modes respectively. Since, I did not wish to redesign Catalyst UDAF structure only for Hive, I have let the original calls and buffer be as they are for PARTIAL1/COMPLETE mode and have created the `partial2ModeBuffer` exclusively for PARTIAL2/FINAL mode operations. Thus, to answer your question, `buffer` here is used for Partial1 mode operations and `partial2ModeBuffer` is used for Partial2 mode operations respectively. I hope that answers your question. Thank you once again for reviewing @cloud-fan .

Reply:
But I don't quite understand how you make the Hive UDAF work with Spark's two phase aggregate?

Reply:
I see. Can you please elaborate more on the two phase aggregate functionality by Spark? That will help me understand and answer your question better. Thank you.

Reply:
Let's start with the 5 phases of a UDAF:
- **Initialize:** The aggregation buffers for PARTIAL1 Mode and PARTIAL2 Mode are created in this phase.
- **Iterate(Update) :** This state processes a new row of data into the aggregation buffer created for PARTIAL1.
- **TerminatePartial:** Returns the contents of the aggregation buffer.
- **Merge:** Merges a partial aggregation returned by calling terminatePartial() on PARTIAL1 aggregation buffer into the current aggregation happening on PARTIAL2 aggregation buffer.
- **Terminate:** Returns the final result of the aggregation stored in PARTIAL2 buffer to Hive.

In Spark, a UDAF will be run twice in two adjacent aggregate operators, called partial aggregate and final aggregate. In the partial aggregate, there are 3 steps:
1. initialize the UDAF
2. update UDAF with input data (so-called Iterate)
3. return the UDAF buffer (so-called TerminatePartial)

In the final aggregate, also 3 steps:
1. initialize the UDAF
2. update UDAF with buffer data from the partial aggregate (so-called Merge)
3. return final result (so-called Terminate)

But this doesn't work for the 3-phase UDAF which doesn't support partial aggregate.

Reply:
Ah good point. I did come across this while I was designing the code. I could see that either of the two (initialize, update, terminate) or (initialize, merge, terminate) steps would run, sometimes one after the other and so on. My fix works for both types of UDAF's(those that support partial aggregate and those that not).

 For the case, when partial aggregate(init, update, terminate) is called, the partial2 aggregate buffer maintains an exact copy of partial1 buffer(https://github.com/apache/spark/pull/23778/files#diff-9463c978126246071e528ddfa7a376d5R651). Thus, for the above case, partial2 buffer is the same as partial1. However for the case, when final aggregate(init, merge, terminate) is called, the partial2 buffer will hold the results of the merge from partial1 buffer and will be used in terminate(eval) to give out the final results. 

In both the cases, partial2 buffer is initialized to hold a copy of partial1 buffer data here(https://github.com/apache/spark/pull/23778/files#diff-9463c978126246071e528ddfa7a376d5R644).

Reply:
even for hive UDAF which doesn't support partial aggregate, it still has partial1 buffer? Can you list the steps of the 2 phase aggregate with the UDAF that doesn't support partial aggregate, like what I did?

Reply:
Sure. Take for an example, the test **"testDISTs"** in `HiveWindowFunctionQuerySuite` where the query `"""
      |select  p_mfgr,p_name, p_size,
      |histogram_numeric(p_retailprice, 5) over w1 as hist,
      |percentile(p_partkey, cast(0.5 as double)) over w1 as per,
      |row_number() over(distribute by p_mfgr sort by p_name) as rn
      |from part
      |window w1 as (distribute by p_mfgr sort by p_mfgr, p_name
      |             rows between 2 preceding and 2 following)
    """.stripMargin` is a Hive UDAF that does not support partial aggregate. The steps here that follow in the execution of this UDAF are: 

**1. Initialize:** initialize the UDAF as well as the buffers for partial1 and partial2.
**2. Update:** update the UDAF with input data. Here, the partial1 buffer holds the result of the update. Each time the update occurs, a copy operation happens where the partial1 buffer is copied to the partial2 buffer.
**3. Eval(Terminate):** return the final result held by the partial2 buffer.

The actual sequence of calls goes like this: Initialize, Update, Update, Update...., Eval, Initialize, Update, Update...., Eval and so on. 

I admit that the part in the update where partial1 buffer is copied into partial2 buffer each time seems a little less efficient to me and can be worked on.

Reply:
In this case, Hive UDAF is used as a window function, so there is no 2-phase aggregation.

Can you show a normal GROUP BY case and list the steps of executing Hive UDAF?

Reply:
@cloud-fan Sorry for the delay in my response. I tried a bunch of things, looked up for instances of hive UDAF's that do not support partial aggregate, however, all the ones that we currently have seem to support partial aggregation. When you mentioned 

> Can you list the steps of the 2 phase aggregate with the UDAF that doesn't support partial aggregate, like what I did?

Did you have any specific Hive UDAF in mind that you know does not support partial aggregate be it built-in or customized? If so, I can try that out and let you know. Thank you.

Reply:
hmm, looking at your PR description, seems the bug is about 2 buffers instead of 2-phase aggregate.

Can you explain more about how the bug happens?
```
com.yahoo.sketches.hive.hll.SketchState cannot be cast to com.yahoo.sketches.hive.hll.UnionState
at com.yahoo.sketches.hive.hll.SketchEvaluator.merge(SketchEvaluator.java:56)
```

Reply:
Sure. The bug mainly stems from the fact that Spark treats aggregation buffers of modes PARTIAL1/COMPLETE and PARTIAL2/FINAL as one whereas, they are supposed to be treated as two different aggregation buffers initialized for two different mode groups.

 I have reproduced the bug in the Unit test `test("customized Hive UDAF2")` written for this PR.

The other part of this PR deals with fixing a separate bug in SortBasedAggregator which involved calling merge() on the aggregate buffer without ever calling initialize(). This issue was not witnessed before as Spark would only create one aggregation buffer for all the modes.

Reply:
In Hive UDAF, when to use which agg buffer? I think this is the most important information to justify your patch. It will be better if you ca point to some Hive doc/code comments.

Reply:
So I went through Hive docs and asked a couple of people around; officially, hive does not mention anything about using two different aggregation buffers, the main point is to have some kind of distinction between different phases of Hive. 

Consider a classic map-reduce process. There are two phases: map and reduce (sometimes an optional combine phase in between). The phases can run on different nodes. The state lives within a phase and does not cross the boundaries. The map phase corresponds to the "partial1" mode (init + iterate + terminate partial). The reduce phase corresponds to the "final" mode (init + merge + terminate). The combine phase corresponds to the "partial2" mode (init + merge + terminate partial). The "complete" mode is a special shortcut to run the whole thing as a single phase (init + iterate + terminate). The bug here is about a state crossing the boundaries between the phases: initialized for one phase (mode), but then passed to a different phase. So by using different aggregation buffers, I am trying to encapsulate the corresponding state within a particular phase. The solution can also be modified to have a single aggregation buffer supporting states of different phases. 

In my PR above, the assumption is that the Partial1 aggregation buffer supports phases PARTIAL1/COMPLETE and the Partial2 aggregation buffer supports phases PARTIAL2/FINAL. 

I shall also paste a link to a good blog that explains the usage of aggregation buffers in a generic Hive UDAF : https://blog.dataiku.com/2013/05/01/a-complete-guide-to-writing-hive-udf 

As this is also a kind of a design change problem, it is completely open to further discussions and improvements. My solution is just one of a kind solution and there are multiple solutions to achieve the same thing. However, as far as I can say, my solution is relatively cleaner and easier to understand and also it does not create a change of any manner in the way with which existing aggregation functions work with Spark SQL(does not break compatibility).

Reply:
So there are 4 ways to execute a UDAF
1. init + iterate + terminate partial
2. init + merge + terminate final
3. init + merge + terminate partial
4. init + iterate + terminate final

Spark doesn't really have terminate partial. The agg buffer needs to fit the spark schema so Spark can get agg buffer directly. Spark UDAF is flexible: after initialized, the buffer can be updated via either iterate or merge, the buffer can be terminated always.

IIUC `init + merge + terminate final` is pretty common in GROUP BY queries, and Hive UDAF works in this case. Do you know why?

And your test case is `init + iterate + terminate final`, what's the correct steps to do it? Is it
```
1. create partial1 buffer
2. iterate
3 turn partial1 buffer to partial2 buffer
4. terminate final
```

Reply:
@cloud-fan Sorry for the delayed response, I agree with your point above. However, I did not understand the question correctly.

> IIUC init + merge + terminate final is pretty common in GROUP BY queries, and Hive UDAF works in this case. Do you know why?

> And your test case is init + iterate + terminate final
Actually my test case is `init + iterate + terminate partial` alongwith `init + iterate + merge + terminate partial` and finally ending with `init + merge + terminate final`. So according to me, the correct steps here would roughly be:

```
1. create partial1 buffer
2. iterate
3. merge partial1 buffer into partial2 buffer
4. terminate final
```

Apologies if I have misread your above comment and have not answered it appropriately, please let me know. Thank you.

Reply:
to make it clear, the partial 1 buffer can only be used in `iterator` to consume records, and partial 2 buffer can only be used in `merge` to consume buffers, do I understand it correctly?